### PR TITLE
MDEV-19698 - remove dead code

### PR DESCRIPTION
**Pull request by kevgs**
I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

### CODE REVIEW

File: sql/records.cc [hunk 1]
```diff
@@ -664,8 +663,7 @@ static int rr_from_cache(READ_RECORD *info)
       int3store(ref_position,(long) i);
       ref_position+=3;
     }
-    my_qsort(info->read_positions, length, info->struct_length,
-             (qsort_cmp) rr_cmp);
+    my_qsort(info->read_positions, length, STRUCT_LENGTH, (qsort_cmp)rr_cmp);
```
Comment (svoj) [inline sql/records.cc]: According to our coding style there should be space between cast and variable: `(qsort_cmp) rr_cmp)`.

File: sql/records.h [hunk 2]
```diff
@@ -71,6 +66,7 @@ struct READ_RECORD
   void    (*unpack)(struct st_sort_addon_field *, uchar *, uchar *);
 
   int read_record() { return read_record_func(this); }
+  uchar *record();
```
Comment (svoj) [inline sql/records.h]: It should be inline. We don't really want to waste time for function call convention here.

**Comment by svoj** [comment]
@kevgs I could cherry-pick some of the changes which I agree with, or would you prefer to have this PR merged as a whole after discussing individual things?

File: sql/records.cc (hunk 1 above)
Comment (kevgs) [reply sql/records.cc]: Is coding style written on some web page?

**Comment by kevgs** [comment]
One commit is better.

You can use github feature called `Squash and merge` while applying this PR to `10.3`. All commits from PR will be squashed to one commit. You may also fix commit message. Is it ok or I my squash commits in PR by myself after review.

**Comment by svoj** [comment]
Ok, I'll try to make another review round soon. Although I'd probably keep some changes in separate commits.

**Comment by kevgs** [comment]
> Looks like MySQL removed coding style guidelines that we've been using.

They have [this](https://github.com/mysql/mysql-server/blob/8.0/.clang-format)

**Comment by svoj** [comment]
Sorry for a long delay. I've applied most of the patches. Could you rebase the rest so that we shrink scope of the discussion?

File: storage/innobase/btr/btr0cur.cc [hunk 3]
```diff
@@ -4605,10 +4605,6 @@ btr_cur_pessimistic_update(
 
 	rec = btr_cur_get_rec(cursor);
 
-	*offsets = rec_get_offsets(
-		rec, index, *offsets, page_is_leaf(page),
-		ULINT_UNDEFINED, offsets_heap);
-
```
Comment (dr-m) [inline storage/innobase/btr/btr0cur.cc]: Good catch! This applies to 10.2 already. But, please let us assert that it is valid:
```c++
ut_ad(rec_offs_validate(rec, index, *offsets));
```

File: storage/innobase/handler/handler0alter.cc [hunk 4]
```diff
@@ -5969,192 +5952,6 @@ innobase_check_foreign_key_index(
...
-		kick-in in a repeated ALTER sequences where the old name is
-		alternately longer/shorter than the new name:
-		1. ALTER TABLE t RENAME INDEX a TO aa;
-		2. ALTER TABLE t RENAME INDEX aa TO a;
-		3. go to 1. */
-		index->name = mem_heap_strdup_replace(
-			index->heap,
-			/* Presumed topmost element of the heap: */
-			index->name, old_name_len + 1,
-			new_name);
```
Comment (dr-m) [inline storage/innobase/handler/handler0alter.cc]: The now-dead functions `mem_heap_strdup_replace()`, `mem_heap_dup_replace()`, `mem_heap_replace()` and `mem_heap_is_top()` should be removed as well.

**Comment by kevgs** [comment]
@svoj PR was updated.

**Comment by svoj** [comment]
I split your patch and pushed some parts of it with minor edits to 10.3. We should probably keep `bits_set()` as it is being used in 10.4.

I don't feel confident about `READ_RECORD::record` removal. Could you update this PR once again and we assign another reviewer for this particular change.

Patch:
```patch
commit 2221d8505c79baa4f3d9010256327083d1bcff37
Author: Eugene Kosov <claprix@yandex.ru>
Date:   Tue May 29 23:25:50 2018 +0300

    cleanup READ_RECORD

diff --git a/sql/item_subselect.cc b/sql/item_subselect.cc
index 538b2c2..4603a85 100644
--- a/sql/item_subselect.cc
+++ b/sql/item_subselect.cc
@@ -3877,7 +3877,6 @@ int subselect_single_select_engine::exec()
               tab->save_read_record= tab->read_record.read_record_func;
               tab->read_record.read_record_func= rr_sequential;
               tab->read_first_record= read_first_record_seq;
-              tab->read_record.record= tab->table->record[0];
               tab->read_record.thd= join->thd;
               tab->read_record.ref_length= tab->table->file->ref_length;
               tab->read_record.unlock_row= rr_unlock_row;
@@ -3895,7 +3894,6 @@ int subselect_single_select_engine::exec()
     for (JOIN_TAB **ptab= changed_tabs; ptab != last_changed_tab; ptab++)
     {
       JOIN_TAB *tab= *ptab;
-      tab->read_record.record= 0;
       tab->read_record.ref_length= 0;
       tab->read_first_record= tab->save_read_first_record;
       tab->read_record.read_record_func= tab->save_read_record;
diff --git a/sql/records.cc b/sql/records.cc
index c2db29c..5dd4318 100644
--- a/sql/records.cc
+++ b/sql/records.cc
@@ -77,7 +77,6 @@ bool init_read_record_idx(READ_RECORD *info, THD *thd, TABLE *table,
   bzero((char*) info,sizeof(*info));
   info->thd= thd;
   info->table= table;
-  info->record= table->record[0];
   info->print_error= print_error;
   info->unlock_row= rr_unlock_row;

@@ -210,7 +209,6 @@ bool init_read_record(READ_RECORD *info,THD *thd, TABLE *table,
   else
   {
     empty_record(table);
-    info->record= table->record[0];
     info->ref_length= (uint)table->file->ref_length;
   }
   info->select=select;
@@ -393,7 +391,7 @@ static int rr_index_first(READ_RECORD *info)
     return tmp;
   }

-  tmp= info->table->file->ha_index_first(info->record);
+  tmp= info->table->file->ha_index_first(info->record());
   info->read_record_func= rr_index;
   if (tmp)
     tmp= rr_handle_error(info, tmp);
@@ -416,7 +414,7 @@ static int rr_index_first(READ_RECORD *info)

 static int rr_index_last(READ_RECORD *info)
 {
-  int tmp= info->table->file->ha_index_last(info->record);
+  int tmp= info->table->file->ha_index_last(info->record());
   info->read_record_func= rr_index_desc;
   if (tmp)
     tmp= rr_handle_error(info, tmp);
@@ -442,7 +440,7 @@ static int rr_index_last(READ_RECORD *info)

 static int rr_index(READ_RECORD *info)
 {
-  int tmp= info->table->file->ha_index_next(info->record);
+  int tmp= info->table->file->ha_index_next(info->record());
   if (tmp)
     tmp= rr_handle_error(info, tmp);
   return tmp;
@@ -467,7 +465,7 @@ static int rr_index(READ_RECORD *info)

 static int rr_index_desc(READ_RECORD *info)
 {
-  int tmp= info->table->file->ha_index_prev(info->record);
+  int tmp= info->table->file->ha_index_prev(info->record());
   if (tmp)
     tmp= rr_handle_error(info, tmp);
   return tmp;
@@ -477,7 +475,7 @@ static int rr_index_desc(READ_RECORD *info)
 int rr_sequential(READ_RECORD *info)
 {
   int tmp;
-  while ((tmp= info->table->file->ha_rnd_next(info->record)))
+  while ((tmp= info->table->file->ha_rnd_next(info->record())))
   {
     tmp= rr_handle_error(info, tmp);
     break;
@@ -493,7 +491,7 @@ static int rr_from_tempfile(READ_RECORD *info)
   {
     if (my_b_read(info->io_cache,info->ref_pos,info->ref_length))
       return -1;					/* End of file */
-    if (!(tmp= info->table->file->ha_rnd_pos(info->record,info->ref_pos)))
+    if (!(tmp= info->table->file->ha_rnd_pos(info->record(), info->ref_pos)))
       break;
     /* The following is extremely unlikely to happen */
     if (tmp == HA_ERR_KEY_NOT_FOUND)
@@ -543,7 +541,7 @@ int rr_from_pointers(READ_RECORD *info)
     cache_pos= info->cache_pos;
     info->cache_pos+= info->ref_length;

-    if (!(tmp= info->table->file->ha_rnd_pos(info->record,cache_pos)))
+    if (!(tmp= info->table->file->ha_rnd_pos(info->record(), cache_pos)))
       break;

     /* The following is extremely unlikely to happen */
@@ -638,7 +636,7 @@ static int rr_from_cache(READ_RECORD *info)
       else
       {
 	error=0;
-	memcpy(info->record,info->cache_pos,
+        memcpy(info->record(), info->cache_pos,
                (size_t) info->table->s->reclength);
       }
       info->cache_pos+=info->reclength;
diff --git a/sql/records.h b/sql/records.h
index e97f6b2..15bb7be 100644
--- a/sql/records.h
+++ b/sql/records.h
@@ -19,9 +19,10 @@
 #pragma interface                      /* gcc class implementation */
 #endif

+#include "table.h"
+
 struct st_join_table;
 class handler;
-struct TABLE;
 class THD;
 class SQL_SELECT;
 class Copy_field;
@@ -58,7 +59,6 @@ struct READ_RECORD
   SQL_SELECT *select;
   uint ref_length, reclength, rec_cache_size, error_offset;
   uchar *ref_pos;				/* pointer to form->refpos */
-  uchar *record;
   uchar *rec_buf;                /* to read field values  after filesort */
   uchar	*cache,*cache_pos,*cache_end,*read_positions;
   struct st_sort_addon_field *addon_field;     /* Pointer to the fields info */
@@ -67,6 +67,7 @@ struct READ_RECORD
   void    (*unpack)(struct st_sort_addon_field *, uchar *, uchar *);

   int read_record() { return read_record_func(this); }
+  uchar *record() { return table->record[0]; }

   /*
     SJ-Materialization runtime may need to read fields from the materialized
diff --git a/sql/sql_select.cc b/sql/sql_select.cc
index b731b18..eb77276 100644
--- a/sql/sql_select.cc
+++ b/sql/sql_select.cc
@@ -20392,7 +20392,6 @@ join_read_first(JOIN_TAB *tab)
   tab->table->status=0;
   tab->read_record.read_record_func= join_read_next;
   tab->read_record.table=table;
-  tab->read_record.record=table->record[0];
   if (!table->file->inited)
     error= table->file->ha_index_init(tab->index, tab->sorted);
   if (likely(!error))
@@ -20412,7 +20411,7 @@ static int
 join_read_next(READ_RECORD *info)
 {
   int error;
-  if (unlikely((error= info->table->file->ha_index_next(info->record))))
+  if (unlikely((error= info->table->file->ha_index_next(info->record()))))
     return report_error(info->table, error);

   return 0;
@@ -20432,7 +20431,6 @@ join_read_last(JOIN_TAB *tab)
   tab->table->status=0;
   tab->read_record.read_record_func= join_read_prev;
   tab->read_record.table=table;
-  tab->read_record.record=table->record[0];
   if (!table->file->inited)
     error= table->file->ha_index_init(tab->index, 1);
   if (likely(!error))
@@ -20449,7 +20447,7 @@ static int
 join_read_prev(READ_RECORD *info)
 {
   int error;
-  if (unlikely((error= info->table->file->ha_index_prev(info->record))))
+  if (unlikely((error= info->table->file->ha_index_prev(info->record()))))
     return report_error(info->table, error);
   return 0;
 }
@@ -20479,7 +20477,7 @@ static int
 join_ft_read_next(READ_RECORD *info)
 {
   int error;
-  if (unlikely((error= info->table->file->ha_ft_read(info->table->record[0]))))
+  if (unlikely((error= info->table->file->ha_ft_read(info->record()))))
     return report_error(info->table, error);
   return 0;
 }
diff --git a/sql/sql_window.cc b/sql/sql_window.cc
index 4872429..a6c9dd3 100644
--- a/sql/sql_window.cc
+++ b/sql/sql_window.cc
@@ -898,7 +898,7 @@ class Table_read_cursor : public Rowid_seq_cursor
   {
     Rowid_seq_cursor::init(info);
     table= info->table;
-    record= info->record;
+    record= info->record();
   }

   virtual int fetch()
```

**Comment by svoj** [comment]
@sanja-byelkin please review (and apply?) the above patch. Other changes suggested in this PR should be disregarded as they have been applied already.

What I'm unsure about:
1. is `READ_RECORD::table` constant after initialisation?
2. why `subselect_single_select_engine::exec()` had to reset `READ_RECORD::table`?

**Comment by sanja-byelkin** [comment]
moved to the sources as 490dcfd5d74ce45cbdd6105e4295116f6a944e2b

**Pull request closed without merging**